base/compiler/typeinfer.jl

```diff
@@ -412,10 +416,21 @@ function finish(me::InferenceState, interp::AbstractInterpreter)
         me.result.src = OptimizationState(me, OptimizationParams(interp), interp)
     end
     me.result.valid_worlds = me.valid_worlds
-    me.result.result = me.bestguess
+    me.result.result = me.bestguess = bestguess_to_interprocedural(me.bestguess, me.nargs)
     nothing
 end
 
+bestguess_to_interprocedural(@nospecialize(bestguess), _::Int) = bestguess
+function bestguess_to_interprocedural(bestguess::Conditional, nargs::Int)
+    # keep `Conditional` return type only when it constrains any of call argument
+    i = slot_id(bestguess.var)
+    if 1 < i ≤ nargs
```

Comment:
I excluded `1` because I couldn't come up with any practical example where we can benefit from constraints on a called generic function itself.
(I actually came up with this one
```julia
struct F{T}
    v::T
end
(f::F)(T) = isa(f, F{T})

code_typed((F,); optimize = false) do f
    f(Int) && return f.v # ::Int
    0
end |> first
``` 
but this kind of code seems really useless)

Well at the same time, I don't think there is any disadvantage to have an `InterConditional` on a slot 1, though.

Reply:
yeah, it's not likely to happen, but seemed feasible to easily include, and that just means it won't typically cost anything